atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/AbortedTransactionSoftCache.java

```diff
@@ -0,0 +1,191 @@
+        PatchyCache refreshedPatchyCache;
+
+        if (maybeSnapshot.isEmpty() || maybeSnapshot.get().bucket < latestBucketSeenSoFar) {
+            refreshedPatchyCache = loadPatchyBucket(latestTsSeenSoFar, latestBucketSeenSoFar);
+        } else {
+            // need to load a range of timestamps
+            refreshedPatchyCache = extendPatch(maybeSnapshot.get(), latestTsSeenSoFar);
+        }
+
+        tryUpdate(refreshedPatchyCache);
+        return refreshedPatchyCache;
+    }
+
+    @Override
+    public void close() {
+        autobatcher.close();
+    }
+
+    private PatchyCache extendPatch(PatchyCache snapshot, long latestTs) {
+        long currentLastKnownConcluded = snapshot.lastKnownConcludedTimestamp;
+
+        // It is possible that cache was update by the time we reach here and refresh is not required any more.
+        if (latestTs <= currentLastKnownConcluded) {
+            return snapshot;
+        }
+
+        Set<Long> newAbortedTransactions =
+                futileTimestampStore.getAbortedTransactionsInRange(currentLastKnownConcluded, latestTs);
+        snapshot.extend(latestTs, newAbortedTransactions);
+        return snapshot;
+    }
+
+    private PatchyCache loadPatchyBucket(long latestTsSeenSoFar, long latestBucketSeenSoFar) {
+        long maxTsInCurrentBucket = Utils.getMaxTsInCurrentBucket(latestBucketSeenSoFar);
+
+        Set<Long> futileTimestamps = futileTimestampStore.getAbortedTransactionsInRange(
+                Utils.getMinTsInBucket(latestBucketSeenSoFar), Utils.getMaxTsInCurrentBucket(latestBucketSeenSoFar));
+
+        return new PatchyCache(Math.min(latestTsSeenSoFar, maxTsInCurrentBucket), futileTimestamps);
+    }
+
+    private Optional<PatchyCache> getSnapshot() {
+        return Optional.ofNullable(patchyCacheRef.get());
+    }
+
+    private void tryUpdate(PatchyCache update) {
+        patchyCacheRef.getAndAccumulate(update, AbortedTransactionSoftCache::getLatest);
+    }
+
+    private TransactionSoftCacheStatus getStatus(long startTimestamp, Set<Long> abortedTransactions) {
+        return abortedTransactions.contains(startTimestamp)
+                ? TransactionSoftCacheStatus.IS_ABORTED
+                : TransactionSoftCacheStatus.IS_NOT_ABORTED;
+    }
+
+    static PatchyCache getLatest(PatchyCache current, PatchyCache update) {
+        if (current == null) {
+            return update;
+        }
+
+        if (current == update) {
+            return current;
+        }
+
+        if (current.bucket == update.bucket) {
+            return current.lastKnownConcludedTimestamp >= update.lastKnownConcludedTimestamp ? current : update;
+        }
+
+        return current.bucket > update.bucket ? current : update;
+    }
+
+    static class PatchyCache {
+        private final long bucket;
+        private final Set<Long> abortedTransactions;
+        private long lastKnownConcludedTimestamp;
```

Comment:
Does this need to be `volatile` since this may be read/written by multiple threads?

Reply:
I think the volatile read of the patchy cache should make it so that readers do see the latest value? Since access is serial and it's always volatile read, read, write